packages/plugin-compiler/src/main/kotlin/io/realm/compiler/AccessorModifierIrGeneration.kt

```diff
@@ -430,36 +430,53 @@ class AccessorModifierIrGeneration(private val pluginContext: IrPluginContext) {
         return propertyClassId == realmListClassId
     }
 
-    private fun getListGenericType(declaration: IrProperty): List<String> {
+    private fun getListGenericCoreType(declaration: IrProperty): CoreType {
         // Check first if the generic is a subclass of RealmObject
-        val superTypes = declaration.symbol.descriptor.type.arguments[0].type.constructor.supertypes
-        for (superType: KotlinType in superTypes) {
+        val descriptorType = declaration.symbol.descriptor.type
+        val listGenericType = descriptorType.arguments[0].type
+        for (superType in listGenericType.constructor.supertypes) {
             if (superType.toString() == "RealmObject") {
-                return listOf("RealmObject")
+                // Nullable objects are not supported
+                if (listGenericType.isNullable()) {
+                    error("Error in field ${declaration.name} - RealmLists can only contain non-nullable RealmObjects.")
+                }
+                return CoreType(
+                    propertyType = PropertyType.RLM_PROPERTY_TYPE_OBJECT,
+                    nullable = false
+                )
             }
         }
 
-        // Otherwise just return the generic(s) present in the declaration
-        return declaration.symbol.descriptor.type.arguments.map {
-            it.toString()
+        // If not a RealmObject, check whether the list itself is nullable - if so, throw error
+        if (descriptorType.isNullable()) {
+            error("Error in field ${declaration.name} - a RealmList field cannot be marked as nullable.")
```

Comment:
I assume we have tests for these use cases?

Reply:
So far I've only tested it manually, but I'm currently working on how to write compiler tests for RealmLists. The only place where we test compiler stuff now is `GenerationExtensionTest` and the feeling I got is that everything would look all over the place if I put list-based tests in that file too. I created https://github.com/realm/realm-kotlin/issues/290 to address this.

Reply:
There are also compiler tests in `test/jvm/kotlin/io/realm/test/compiler`.